### PR TITLE
Preserve default spool.

### PR DESCRIPTION
**Pull request by 3Hren**
### Added
New method: `async_spool`.

    Tries to spool an application non-blocking way (if possible).

### CODE REVIEW

**Comment by 3Hren** [comment]
@diunko @antmat @noxiouz take a look.

File: include/cocaine/api/isolate.hpp [hunk 1]
```diff
@@ -67,8 +68,27 @@ struct isolate_t {
...
+
+        try {
+            spool();
+        } catch(const std::system_error& err) {
+            cb(err.code());
+        } catch(...) {
+            cb(std::make_error_code(std::errc::io_error));
+        }
+
+        cb(std::error_code());
```
Comment (noxiouz) [inline include/cocaine/api/isolate.hpp]: In case of an exception this callback is called twice: with ```err.code()``` and ```0```. Is it ok?

Comment (3Hren) [inline include/cocaine/api/isolate.hpp]: No, good catch.

**Comment by noxiouz** [comment]
LGTM

File: include/cocaine/api/isolate.hpp [hunk 2]
```diff
@@ -67,8 +68,29 @@ struct isolate_t {
...
     virtual
+    void
+    spool() = 0;
+
+    // Default implementation delegates the control flow into the blocking spool method.
+    virtual
     std::unique_ptr<cancellation_t>
-    spool(callback_type cb) = 0;
+    async_spool(callback_type cb) {
+        std::unique_ptr<cancellation_t> cancellation(new cancellation_t());
```
Comment (antmat) [inline include/cocaine/api/isolate.hpp]: What is the purpose of same empty cancellation_t for all? If you want exception safe spool I'd return cancellation_t from private pure virtual spool_impl method and make exception processing here in public spool method. Or if we don't use cancellation at all maybe it's better to drop it completely?

Comment (3Hren) [inline include/cocaine/api/isolate.hpp]: This method is virtual.

Comment (antmat) [inline include/cocaine/api/isolate.hpp]: It looks like exception processing is common and should not be overridden, however cancelation token creation can and should be overridden. Maybe it's better leave cancellation creation in spool? or move it out to some extra method? Do we really need both sync and async implementations for any isolation at the same time? Also async_spool is not a really good name for this method as it has default behaviour which is sync. I'd rather  make it pure virtual with implementation in this case.

Comment (antmat) [inline include/cocaine/api/isolate.hpp]: ok, I got the idea, that it's because of API change. LGTM.

**Pull request merged**